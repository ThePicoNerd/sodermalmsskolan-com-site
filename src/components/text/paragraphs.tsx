import React, { ComponentType, FunctionComponent } from "react";
import { AlertCircle, AlertTriangle, CheckCircle } from "react-feather";
import { HTMLElementProps } from "../../styles/overrides";
import { fonts } from "../../styles/text";

export type ParagraphProps = HTMLElementProps<HTMLParagraphElement>;

export type Paragraph = FunctionComponent<ParagraphProps>;

export const SmallParagraph: Paragraph = (props) => (
  <p
    css={{
      color: "var(--color-text-tertiary)",
      fontSize: "0.875rem",
      margin: 0,
    }}
    {...props}
  />
);

export const CardDescription: Paragraph = (props) => (
  <p
    css={{
      color: "var(--color-text-primary)",
      fontSize: "1rem",
      margin: "1rem 0 0",
      lineHeight: 1.75,
      fontFamily: fonts.sans,
    }}
    {...props}
  />
);

/**
 * Paragraph with an informative icon.
 *
 * @param {any} props Props.
 *
 * @returns {React.ReactElement} The rendered paragraph.
 */
export const IconParagraph: FunctionComponent<ParagraphProps & {
  Icon: ComponentType,
}> = ({ Icon, children, ...props }) => (
  <div
    css={{
      display: "flex",
      margin: "1em 0",
      position: "relative",
      lineHeight: 1.25,
      "--icon-size": "1.5rem",

      "&::after": {
        "--border-width": "2px",
        content: "\"\"",
        position: "absolute",
        width: "var(--border-width)",
        backgroundColor: "currentColor",
        bottom: 0,
        top: "calc(var(--icon-size) + 0.5rem)",
        left: "calc((var(--icon-size) - var(--border-width)) / 2)",
      },
    }}
    {...props}
  >
    <Icon css={{
      marginRight: "0.25rem",
      flex: "0 0 var(--icon-size)",
      height: "var(--icon-size)",
    }}
    />
    <p css={{
      margin: "0.125rem 0",
      fontSize: "1rem",
      fontFamily: fonts.sans,

      a: {
        color: "inherit",
        fontWeight: 500,
      },
    }}
    >
      {children}
    </p>
  </div>
);

export const DangerParagraph: Paragraph = (props) => (
  <IconParagraph
    Icon={AlertCircle}
    css={{
      color: "var(--color-text-danger)",
    }}
    {...props}
  />
);

export const WarningParagraph: Paragraph = (props) => (
  <IconParagraph
    Icon={AlertTriangle}
    css={{
      color: "var(--color-text-warning)",
    }}
    {...props}
  />
);

export const SuccessParagraph: Paragraph = (props) => (
  <IconParagraph
    Icon={CheckCircle}
    css={{
      color: "var(--color-text-success)",
    }}
    {...props}
  />
);
