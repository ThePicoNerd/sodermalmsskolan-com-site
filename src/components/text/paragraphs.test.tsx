import renderer from "react-test-renderer";
import React from "react";
import { Info } from "react-feather";
import { IconParagraph } from "./paragraphs";

describe("icon paragraph", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<IconParagraph icon={Info}>Sample text</IconParagraph>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
