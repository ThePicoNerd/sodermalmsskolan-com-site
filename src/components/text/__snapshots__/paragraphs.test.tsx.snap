// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`icon paragraph renders correctly 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 1em 0;
  position: relative;
  line-height: 1.25;
  --icon-size: 1.5rem;
}

.emotion-0::after {
  --border-width: 2px;
  content: "";
  position: absolute;
  width: var(--border-width);
  background-color: currentColor;
  bottom: 0;
  top: calc(var(--icon-size) + 0.5rem);
  left: calc((var(--icon-size) - var(--border-width)) / 2);
}

.emotion-1 {
  margin-right: 0.25rem;
  -webkit-flex: 0 0 var(--icon-size);
  -ms-flex: 0 0 var(--icon-size);
  flex: 0 0 var(--icon-size);
  height: var(--icon-size);
}

.emotion-2 {
  margin: 0.125rem 0;
  font-size: 1rem;
  font-family: Inter,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen,Ubuntu,Cantarell,"Open Sans","Helvetica Neue",Helvetica,sans-serif;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-2 a {
  color: inherit;
  font-weight: 500;
}

<div
  className="emotion-0"
>
  <svg
    className="emotion-1"
    fill="none"
    height={24}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth="2"
    viewBox="0 0 24 24"
    width={24}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="12"
      cy="12"
      r="10"
    />
    <line
      x1="12"
      x2="12"
      y1="16"
      y2="12"
    />
    <line
      x1="12"
      x2="12.01"
      y1="8"
      y2="8"
    />
  </svg>
  <p
    className="emotion-2"
  >
    Sample text
  </p>
</div>
`;
